[dynamo, hops] refactor HOP graph breaks to always include HOP names

williamwen42 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #170115
* #170031
* __->__ #169742

Fixes most of the rest of https://github.com/pytorch/pytorch/issues/168206.

Graph break error messages now automatically show which HOP is being traced, making debugging easier.
  - HOPs' `call_function` now adds HOP name via `_hop_name` attribute directly added to any caught `Unsupported` exceptions
  - Standardized all HOP names to fully qualified format (e.g., `torch.cond`, `torch.ops.higher_order.map`)

Sample logging from the repro in the linked issue:
```
Traceback (most recent call last):
  File "/data/users/williamwen/pytorch/playground.py", line 71, in <module>
    print(opt_mod(x))
          ^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/_dynamo/eval_frame.py", line 441, in __call__
    return super().__call__(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1776, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1787, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/_dynamo/eval_frame.py", line 936, in compile_wrapper
    raise e.with_traceback(None) from e.__cause__  # User compiler error
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
torch._dynamo.exc.Unsupported: HOP: Unsafe side effect
  Higher Order Operator: torch.utils.checkpoint.checkpoint
  Explanation: Mutating a variable from outside the scope of this HOP is not supported.
  Hint: If the HOP is activation checkpointing, try setting `torch._dynamo.config.skip_fwd_side_effects_in_bwd_under_checkpoint = True`

  Developer debug context: Attempted to mutate ListVariable(length=1)

 For more details about this graph break, please visit: https://meta-pytorch.github.io/compile-graph-break-site/gb/gb0067.html

from user code:
   File "/data/users/williamwen/pytorch/playground.py", line 17, in wrapped
    out = module_or_fn(*args, **kwargs)
  File "/data/users/williamwen/pytorch/playground.py", line 61, in forward
    return torch.cos(x) + self.submod(x)[0]
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1776, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1787, in _call_impl
    return forward_call(*args, **kwargs)
  File "/data/users/williamwen/pytorch/playground.py", line 16, in wrapped
    with CtxManager() as ctx:
  File "/data/users/williamwen/pytorch/playground.py", line 25, in __enter__
    stack.append(1)

Set TORCHDYNAMO_VERBOSE=1 for the internal stack trace (please do this especially if you're reporting a bug to PyTorch). For even more developer context, set TORCH_LOGS="+dynamo"
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @Lucaskabela

williamwen42 committed:
[dynamo, hops] refactor HOP graph breaks to always include HOP names

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, hops] refactor HOP graph breaks to always include HOP names"


Fixes most of the rest of https://github.com/pytorch/pytorch/issues/168206.

Graph break error messages now automatically show which HOP is being traced, making debugging easier.
  - Added automatic HOP detection via `_get_current_hop_context()` in exc.py
  - Enhanced `raise_hard_error_if_graph_break` decorator to include HOP names for ObservedException cases
  - Standardized all HOP names to fully qualified format (e.g., `torch.cond`, `torch.ops.higher_order.map`)

Sample logging from the repro in the linked issue:
```
Traceback (most recent call last):
  File "/data/users/williamwen/pytorch/playground.py", line 71, in <module>
    print(opt_mod(x))
          ^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/_dynamo/eval_frame.py", line 441, in __call__
    return super().__call__(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1776, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1787, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/_dynamo/eval_frame.py", line 936, in compile_wrapper
    raise e.with_traceback(None) from e.__cause__  # User compiler error
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
torch._dynamo.exc.Unsupported: HigherOrderOperator: Mutating a variable not in the current scope (SideEffects)
  Explanation: Mutating a variable from outside the scope of the HOP is not supported.
  Hint: If the HOP is activation checkpointing, try setting `torch._dynamo.config.skip_fwd_side_effects_in_bwd_under_checkpoint = True`
  Higher Order Operator: torch.utils.checkpoint.checkpoint

  Developer debug context: Attempted to mutate ListVariable(length=1)

 For more details about this graph break, please visit: https://meta-pytorch.github.io/compile-graph-break-site/gb/gb0067.html

from user code:
   File "/data/users/williamwen/pytorch/playground.py", line 17, in wrapped
    out = module_or_fn(*args, **kwargs)
  File "/data/users/williamwen/pytorch/playground.py", line 61, in forward
    return torch.cos(x) + self.submod(x)[0]
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1776, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1787, in _call_impl
    return forward_call(*args, **kwargs)
  File "/data/users/williamwen/pytorch/playground.py", line 16, in wrapped
    with CtxManager() as ctx:
  File "/data/users/williamwen/pytorch/playground.py", line 25, in __enter__
    stack.append(1)

Set TORCHDYNAMO_VERBOSE=1 for the internal stack trace (please do this especially if you're reporting a bug to PyTorch). For even more developer context, set TORCH_LOGS="+dynamo"
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, hops] refactor HOP graph breaks to always include HOP names"


Fixes most of the rest of https://github.com/pytorch/pytorch/issues/168206.

Graph break error messages now automatically show which HOP is being traced, making debugging easier.
  - Added automatic HOP detection via `_get_current_hop_context()` in exc.py
  - Enhanced `raise_hard_error_if_graph_break` decorator to include HOP names for ObservedException cases
  - Standardized all HOP names to fully qualified format (e.g., `torch.cond`, `torch.ops.higher_order.map`)

Sample logging from the repro in the linked issue:
```
Traceback (most recent call last):
  File "/data/users/williamwen/pytorch/playground.py", line 71, in <module>
    print(opt_mod(x))
          ^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/_dynamo/eval_frame.py", line 441, in __call__
    return super().__call__(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1776, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1787, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/_dynamo/eval_frame.py", line 936, in compile_wrapper
    raise e.with_traceback(None) from e.__cause__  # User compiler error
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
torch._dynamo.exc.Unsupported: HigherOrderOperator: Mutating a variable not in the current scope (SideEffects)
  Explanation: Mutating a variable from outside the scope of the HOP is not supported.
  Hint: If the HOP is activation checkpointing, try setting `torch._dynamo.config.skip_fwd_side_effects_in_bwd_under_checkpoint = True`
  Higher Order Operator: torch.utils.checkpoint.checkpoint

  Developer debug context: Attempted to mutate ListVariable(length=1)

 For more details about this graph break, please visit: https://meta-pytorch.github.io/compile-graph-break-site/gb/gb0067.html

from user code:
   File "/data/users/williamwen/pytorch/playground.py", line 17, in wrapped
    out = module_or_fn(*args, **kwargs)
  File "/data/users/williamwen/pytorch/playground.py", line 61, in forward
    return torch.cos(x) + self.submod(x)[0]
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1776, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1787, in _call_impl
    return forward_call(*args, **kwargs)
  File "/data/users/williamwen/pytorch/playground.py", line 16, in wrapped
    with CtxManager() as ctx:
  File "/data/users/williamwen/pytorch/playground.py", line 25, in __enter__
    stack.append(1)

Set TORCHDYNAMO_VERBOSE=1 for the internal stack trace (please do this especially if you're reporting a bug to PyTorch). For even more developer context, set TORCH_LOGS="+dynamo"
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, hops] refactor HOP graph breaks to always include HOP names"


Fixes most of the rest of https://github.com/pytorch/pytorch/issues/168206.

Graph break error messages now automatically show which HOP is being traced, making debugging easier.
  - Added automatic HOP detection via `_get_current_hop_context()` in exc.py
  - Enhanced `raise_hard_error_if_graph_break` decorator to include HOP names for ObservedException cases
  - Standardized all HOP names to fully qualified format (e.g., `torch.cond`, `torch.ops.higher_order.map`)

Sample logging from the repro in the linked issue:
```
Traceback (most recent call last):
  File "/data/users/williamwen/pytorch/playground.py", line 71, in <module>
    print(opt_mod(x))
          ^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/_dynamo/eval_frame.py", line 441, in __call__
    return super().__call__(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1776, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1787, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/_dynamo/eval_frame.py", line 936, in compile_wrapper
    raise e.with_traceback(None) from e.__cause__  # User compiler error
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
torch._dynamo.exc.Unsupported: HigherOrderOperator: Mutating a variable not in the current scope (SideEffects)
  Explanation: Mutating a variable from outside the scope of the HOP is not supported.
  Hint: If the HOP is activation checkpointing, try setting `torch._dynamo.config.skip_fwd_side_effects_in_bwd_under_checkpoint = True`
  Higher Order Operator: torch.utils.checkpoint.checkpoint

  Developer debug context: Attempted to mutate ListVariable(length=1)

 For more details about this graph break, please visit: https://meta-pytorch.github.io/compile-graph-break-site/gb/gb0067.html

from user code:
   File "/data/users/williamwen/pytorch/playground.py", line 17, in wrapped
    out = module_or_fn(*args, **kwargs)
  File "/data/users/williamwen/pytorch/playground.py", line 61, in forward
    return torch.cos(x) + self.submod(x)[0]
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1776, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1787, in _call_impl
    return forward_call(*args, **kwargs)
  File "/data/users/williamwen/pytorch/playground.py", line 16, in wrapped
    with CtxManager() as ctx:
  File "/data/users/williamwen/pytorch/playground.py", line 25, in __enter__
    stack.append(1)

Set TORCHDYNAMO_VERBOSE=1 for the internal stack trace (please do this especially if you're reporting a bug to PyTorch). For even more developer context, set TORCH_LOGS="+dynamo"
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, hops] refactor HOP graph breaks to always include HOP names"


Fixes most of the rest of https://github.com/pytorch/pytorch/issues/168206.

Graph break error messages now automatically show which HOP is being traced, making debugging easier.
  - Added automatic HOP detection via `_get_current_hop_context()` in exc.py
  - Enhanced `raise_hard_error_if_graph_break` decorator to include HOP names for ObservedException cases
  - Standardized all HOP names to fully qualified format (e.g., `torch.cond`, `torch.ops.higher_order.map`)

Sample logging from the repro in the linked issue:
```
Traceback (most recent call last):
  File "/data/users/williamwen/pytorch/playground.py", line 71, in <module>
    print(opt_mod(x))
          ^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/_dynamo/eval_frame.py", line 441, in __call__
    return super().__call__(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1776, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1787, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/_dynamo/eval_frame.py", line 936, in compile_wrapper
    raise e.with_traceback(None) from e.__cause__  # User compiler error
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
torch._dynamo.exc.Unsupported: HigherOrderOperator: Mutating a variable not in the current scope (SideEffects)
  Explanation: Mutating a variable from outside the scope of the HOP is not supported.
  Hint: If the HOP is activation checkpointing, try setting `torch._dynamo.config.skip_fwd_side_effects_in_bwd_under_checkpoint = True`
  Higher Order Operator: torch.utils.checkpoint.checkpoint

  Developer debug context: Attempted to mutate ListVariable(length=1)

 For more details about this graph break, please visit: https://meta-pytorch.github.io/compile-graph-break-site/gb/gb0067.html

from user code:
   File "/data/users/williamwen/pytorch/playground.py", line 17, in wrapped
    out = module_or_fn(*args, **kwargs)
  File "/data/users/williamwen/pytorch/playground.py", line 61, in forward
    return torch.cos(x) + self.submod(x)[0]
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1776, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1787, in _call_impl
    return forward_call(*args, **kwargs)
  File "/data/users/williamwen/pytorch/playground.py", line 16, in wrapped
    with CtxManager() as ctx:
  File "/data/users/williamwen/pytorch/playground.py", line 25, in __enter__
    stack.append(1)

Set TORCHDYNAMO_VERBOSE=1 for the internal stack trace (please do this especially if you're reporting a bug to PyTorch). For even more developer context, set TORCH_LOGS="+dynamo"
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, hops] refactor HOP graph breaks to always include HOP names"


Fixes most of the rest of https://github.com/pytorch/pytorch/issues/168206.

Graph break error messages now automatically show which HOP is being traced, making debugging easier.
  - Added automatic HOP detection via `_get_current_hop_context()` in exc.py
  - Enhanced `raise_hard_error_if_graph_break` decorator to include HOP names for ObservedException cases
  - Standardized all HOP names to fully qualified format (e.g., `torch.cond`, `torch.ops.higher_order.map`)

Sample logging from the repro in the linked issue:
```
Traceback (most recent call last):
  File "/data/users/williamwen/pytorch/playground.py", line 71, in <module>
    print(opt_mod(x))
          ^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/_dynamo/eval_frame.py", line 441, in __call__
    return super().__call__(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1776, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1787, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/_dynamo/eval_frame.py", line 936, in compile_wrapper
    raise e.with_traceback(None) from e.__cause__  # User compiler error
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
torch._dynamo.exc.Unsupported: HigherOrderOperator: Mutating a variable not in the current scope (SideEffects)
  Explanation: Mutating a variable from outside the scope of the HOP is not supported.
  Hint: If the HOP is activation checkpointing, try setting `torch._dynamo.config.skip_fwd_side_effects_in_bwd_under_checkpoint = True`
  Higher Order Operator: torch.utils.checkpoint.checkpoint

  Developer debug context: Attempted to mutate ListVariable(length=1)

 For more details about this graph break, please visit: https://meta-pytorch.github.io/compile-graph-break-site/gb/gb0067.html

from user code:
   File "/data/users/williamwen/pytorch/playground.py", line 17, in wrapped
    out = module_or_fn(*args, **kwargs)
  File "/data/users/williamwen/pytorch/playground.py", line 61, in forward
    return torch.cos(x) + self.submod(x)[0]
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1776, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1787, in _call_impl
    return forward_call(*args, **kwargs)
  File "/data/users/williamwen/pytorch/playground.py", line 16, in wrapped
    with CtxManager() as ctx:
  File "/data/users/williamwen/pytorch/playground.py", line 25, in __enter__
    stack.append(1)

Set TORCHDYNAMO_VERBOSE=1 for the internal stack trace (please do this especially if you're reporting a bug to PyTorch). For even more developer context, set TORCH_LOGS="+dynamo"
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, hops] refactor HOP graph breaks to always include HOP names"


Fixes most of the rest of https://github.com/pytorch/pytorch/issues/168206.

Graph break error messages now automatically show which HOP is being traced, making debugging easier.
  - HOPs' `call_function` now adds HOP name via `_hop_name` attribute directly added to any caught `Unsupported` exceptions
  - Standardized all HOP names to fully qualified format (e.g., `torch.cond`, `torch.ops.higher_order.map`)

Sample logging from the repro in the linked issue:
```
Traceback (most recent call last):
  File "/data/users/williamwen/pytorch/playground.py", line 71, in <module>
    print(opt_mod(x))
          ^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/_dynamo/eval_frame.py", line 441, in __call__
    return super().__call__(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1776, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1787, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/_dynamo/eval_frame.py", line 936, in compile_wrapper
    raise e.with_traceback(None) from e.__cause__  # User compiler error
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
torch._dynamo.exc.Unsupported: HOP: Unsafe side effect
  Higher Order Operator: torch.utils.checkpoint.checkpoint
  Explanation: Mutating a variable from outside the scope of this HOP is not supported.
  Hint: If the HOP is activation checkpointing, try setting `torch._dynamo.config.skip_fwd_side_effects_in_bwd_under_checkpoint = True`

  Developer debug context: Attempted to mutate ListVariable(length=1)

 For more details about this graph break, please visit: https://meta-pytorch.github.io/compile-graph-break-site/gb/gb0067.html

from user code:
   File "/data/users/williamwen/pytorch/playground.py", line 17, in wrapped
    out = module_or_fn(*args, **kwargs)
  File "/data/users/williamwen/pytorch/playground.py", line 61, in forward
    return torch.cos(x) + self.submod(x)[0]
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1776, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1787, in _call_impl
    return forward_call(*args, **kwargs)
  File "/data/users/williamwen/pytorch/playground.py", line 16, in wrapped
    with CtxManager() as ctx:
  File "/data/users/williamwen/pytorch/playground.py", line 25, in __enter__
    stack.append(1)

Set TORCHDYNAMO_VERBOSE=1 for the internal stack trace (please do this especially if you're reporting a bug to PyTorch). For even more developer context, set TORCH_LOGS="+dynamo"
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela

[ghstack-poisoned]

anijain2305 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -3786,7 +3850,11 @@ def _call_function(
+@add_hop_context
 class AutogradFunctionApplyVariable(VariableTracker):
+    _HOP_NAME: str = "autograd.Function"
+    _ALLOW_GRAPH_BREAKS = True
```
Why is this True. Today, HOP = no graph breaks.

anijain2305 commented:
Small requests, but almost done.

williamwen42 committed:
Update on "[dynamo, hops] refactor HOP graph breaks to always include HOP names"


Fixes most of the rest of https://github.com/pytorch/pytorch/issues/168206.

Graph break error messages now automatically show which HOP is being traced, making debugging easier.
  - HOPs' `call_function` now adds HOP name via `_hop_name` attribute directly added to any caught `Unsupported` exceptions
  - Standardized all HOP names to fully qualified format (e.g., `torch.cond`, `torch.ops.higher_order.map`)

Sample logging from the repro in the linked issue:
```
Traceback (most recent call last):
  File "/data/users/williamwen/pytorch/playground.py", line 71, in <module>
    print(opt_mod(x))
          ^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/_dynamo/eval_frame.py", line 441, in __call__
    return super().__call__(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1776, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1787, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/williamwen/pytorch/torch/_dynamo/eval_frame.py", line 936, in compile_wrapper
    raise e.with_traceback(None) from e.__cause__  # User compiler error
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
torch._dynamo.exc.Unsupported: HOP: Unsafe side effect
  Higher Order Operator: torch.utils.checkpoint.checkpoint
  Explanation: Mutating a variable from outside the scope of this HOP is not supported.
  Hint: If the HOP is activation checkpointing, try setting `torch._dynamo.config.skip_fwd_side_effects_in_bwd_under_checkpoint = True`

  Developer debug context: Attempted to mutate ListVariable(length=1)

 For more details about this graph break, please visit: https://meta-pytorch.github.io/compile-graph-break-site/gb/gb0067.html

from user code:
   File "/data/users/williamwen/pytorch/playground.py", line 17, in wrapped
    out = module_or_fn(*args, **kwargs)
  File "/data/users/williamwen/pytorch/playground.py", line 61, in forward
    return torch.cos(x) + self.submod(x)[0]
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1776, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/data/users/williamwen/pytorch/torch/nn/modules/module.py", line 1787, in _call_impl
    return forward_call(*args, **kwargs)
  File "/data/users/williamwen/pytorch/playground.py", line 16, in wrapped
    with CtxManager() as ctx:
  File "/data/users/williamwen/pytorch/playground.py", line 25, in __enter__
    stack.append(1)

Set TORCHDYNAMO_VERBOSE=1 for the internal stack trace (please do this especially if you're reporting a bug to PyTorch). For even more developer context, set TORCH_LOGS="+dynamo"
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela

[ghstack-poisoned]
